_posts/2023-11-30-prevent-broker-scale-down-if-containing-paritition-replicas.md

```diff
@@ -0,0 +1,162 @@
+```shell
+Topic: my-topic	TopicId: VcdsMY9gR1STjURMGEHrlA	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1
+	Topic: my-topic	Partition: 0	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
+	Topic: my-topic	Partition: 1	Leader: 0	Replicas: 0,2,3	Isr: 0,2,3
+	Topic: my-topic	Partition: 2	Leader: 2	Replicas: 2,3,1	Isr: 2,3,1
+```
+
+### Scaling down the replicas (without moving replicas from broker to be removed)
+
+Let's try to scale down the no. of brokers from 4 to 3 while all brokers are currently having partition replicas assigned to them
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    version: 3.6.0
+    replicas: 3            // Changes the replicas to 3
+    listeners:
+# ....    
+```
+
+Now you can apply the updated Kafka custom resource and check if the broker are scaled down or not.
+
+You check the logs using this command:
+```sh
+kubect logs kubectl logs strimzi-cluster-operator-56fb857f7c-9hq6l -n myproject 
+```
+
+Since we didn't move the replicas from the broker to be removed, the scale down will fail and, you will be able to see these errors in the logs
+```shell
+2023-11-30 11:08:12 WARN  AbstractOperator:557 - Reconciliation #150(watch) Kafka(myproject/my-cluster): Failed to reconcile
+io.strimzi.operator.common.model.InvalidResourceException: Cannot scale down brokers [3] because brokers [3] are not empty
+	at io.strimzi.operator.cluster.operator.assembly.KafkaReconciler.lambda$brokerScaleDownCheck$26(KafkaReconciler.java:300) ~[io.strimzi.cluster-operator-0.38.0.jar:0.38.0]
+	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:38) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
+	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
+	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
+	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
+	at io.vertx.core.Promise.complete(Promise.java:66) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
+```
+So these logs are basically telling you that the broker are not empty and therefore the reconciliation is failing and cluster operations are blocked. To get rid of this error you can revert the replica change in the Kafka resource.
+
+### Scaling down the replicas (after emptying partition replicas from broker to be removed)
+
+Let's try to scale down the broker now after emptying partition replicas from the broker to be removed.
+
+We can make use of the `KafkaRebalance` resource in Strimzi with `remove-broker` node configuration for this job. 
+Doing this will make Cruise Control do all the job of rebalancing and moving the partition replicas from the broker that is going to be removed.
```

Comment:
```suggestion
Doing this will allow Cruise Control to handle the task of rebalancing and moving the partition replicas from the broker that is going to be removed.
```